python/recipe.hpp

```diff
@@ -0,0 +1,172 @@
+#pragma once
+
+#include <pybind11/pybind11.h>
+#include <pybind11/pytypes.h>
+#include <pybind11/stl.h>
+
+#include <cell.hpp>
+#include <recipe.hpp>
+#include <rss_cell.hpp>
+#include <lif_cell_description.hpp>
+
+#include "event_generator.hpp"
+
+namespace arb {
+
+namespace py {
+
+// py::recipe is the recipe interface that used by Python.
+// Calls that return generic types return pybind11::object, to avoid
+// having to wrap some C++ types used by the C++ interface (specifically
+// util::unique_any, util::any, std::unique_ptr, etc.)
+// For example, requests for cell description return pybind11::object, instead
+// of util::unique_any used by the C++ recipe interface. The py_recipe_shim defined
+// below can unwrap.
+class recipe {
+public:
+    recipe() = default;
+    virtual ~recipe() {}
+
+    virtual cell_size_type   num_cells() const = 0;
+    virtual pybind11::object cell_description(cell_gid_type gid) const = 0;
+    virtual cell_kind        kind(cell_gid_type gid) const = 0;
+    virtual std::vector<arb::cell_connection> connections_on(cell_gid_type gid) const {
+        return {};
+    };
+    virtual cell_size_type num_sources(cell_gid_type) const {
+        return 0;
+    };
+    virtual cell_size_type num_targets(cell_gid_type) const {
+        return 0;
+    };
+    virtual std::vector<pybind11::object> event_generators(cell_gid_type gid) const {
+        auto guard = pybind11::gil_scoped_acquire();
+        return {};
+    };
+};
+
+class recipe_trampoline: public recipe {
```

Comment:
So this is the shim to call py from C, right?

Reply:
I'm having trouble understanding why there needs to be two shims -- this one and the one labeled shim which then calls this one.

Reply:
No -- I think I see, this is the shim for python to C, and the one labeled "shim" goes the other way, but they both end up calling python.